### PR TITLE
Makes Species datum Non-Global

### PR DESCRIPTION
**Pull request by Fox-McCloud**
What a massive pain in the ass.

@Aurorablade sorry to steal your thunder, not trying to. It's just a project like this I can't really work through and review by myself and pinch and poke at; it's something I have to work through myself.

Here it is, the fully functional and working non-global species datums. This not only makes the species datums non-global, but makes it held on the mob's DNA (or the organ's DNA).

Code Changes
- Species datums are now held on a human mob's DNA
- Each human mob's DNA has a unique species datum instance, so you can alter it without it impacting every other of that species type on the map
- Organs no longer explicitly have a species datum, but instead rely on their own DNA datum species holder var (This worked out very nicely
- Refactored a lot of explicit species checks to be datum based
  - There's some species checks that aren't datum based yet; these are mostly for game modes where the user's prefs (which stores species as a name string) are checked against a list. Don't want to touch these in this PR

Fixes:
- Fixes monkey HUDs not showing up
- Fixes going from monkey to human and not having a HUD
- Fixes some items unequipping when you change species
- Fixes mitocholide not rejuvinating some organs
- Fixes shadowlings not being forced to hatch after a time
- Fixes abductor console not properly clearing some action buttons
- Fixes Vox name generation being based on incorrect syllables
- Consolidates Diona name generation instead of it randomly generating a name in player select and then asking you again, on spawn with a different set of names (the lists have been combined together
- Fixes DNA scrambler so it generates an appropriate species name
- Fixes runtimes with spawning in `/mob/living/carbon/human/interactive`
- Fixes a runtime with apply DNA to a head already deleted
  - This was occurring because of Vox player preview during start up; it was caused by spawns in `set_species` and some organ icon updates
- Fixes Vox lungs going nectrotic/being removed and causing permanent "low on nitrogen" warnings.
- Fixes operatives being bald most of the time

Additions
- Having a robotic head will now allow you to beep, buzz, ping, and the likes
- Claw sharpening actually works
- Abductors have their own headsets
- Can put monkeys in the gibber
- Different species now produce different hide types when put into the gibber (monkies will produce monkey hide, Unathi will produce lizard hide, Diona will produce sheets of wood, and other similar things; not every species has a hide).

Removal
- Removes ability to pick name as a diona nymph when evolving (consistent with ghost roles)

Thus far, no runtimes, icons works, and everything seems a-ok.

:cl: Fox McCloud
fix: Fixes monkeys not having their own HUD
fix: Fixes  monkey to human transformations missing their HUD
fix: Fixes changing species unequipping items
fix: Fixes organs not being rejuvinated properly by mitocholide
fix: Fixes shadowlings not being forced to hatch
add: Having a robotic head will allow you to use buzz, ping, beep, and the likes
tweak: Claw sharpening now works
add: Abductors have their own headset instead of a syndicate one
tweak: Can put monkeys in the gibber
add: gibbing some species will now generate different types of hides (monkeys give monkey hides, unathi give lizard hides, the likes)
fix: fixes Vox name generation based on the wrong syllables
tweak: consolidates diona name lists; you'll no longer be asked to rename your Diona when you start the shift
del: botany grown diona can no longer rename themselves (as is consistent with ghost roles)
fix: Fixes DNA scrambler so it gives a proper randomized species name
fix: Fixed an edge case were Vox would have a permanent nitrogen alert
fix: Fixes operatives being bald most of the time
/:cl:

### CODE REVIEW

File: code/game/dna/dna2.dm [hunk 1]
```diff
@@ -421,7 +420,7 @@ var/global/list/bad_blocks[0]
 	data["UE"] = unique_enzymes
 	data["SE"] = SE.Copy() // This is probably too lazy for my own good
 	data["UI"] = UI.Copy()
-	data["species"] = species // This works because `species` is a string, not a datum
+	data["species"] = species.type // This works because `species` is a string, not a datum
```
Comment (Crazylemon64) [inline code/game/dna/dna2.dm]: You might want to update this comment, it's dangerously misleading now

File: code/modules/mob/living/carbon/human/appearance.dm [hunk 2]
```diff
@@ -3,20 +3,20 @@
 	AC.flags = flags
 	AC.ui_interact(user, state = state)
 
-/mob/living/carbon/human/proc/change_species(var/new_species)
-	if(!new_species || species == new_species || !(new_species in all_species))
-		return
+/mob/living/carbon/human/proc/change_species(datum/species/new_species)
+	if(!new_species || dna.species.type == new_species || !(initial(new_species.name) in all_species))
```
Comment (Fox-McCloud) [inline code/modules/mob/living/carbon/human/appearance.dm]: `!(initial(new_species.name) in all_species)` is probably a dumb check, to be honest.

File: code/modules/surgery/organs/organ.dm [hunk 3]
```diff
@@ -62,8 +59,7 @@
 					blood_DNA = list()
 				blood_DNA[dna.unique_enzymes] = dna.b_type
 	else
-		if(istext(species))
-			species = all_species[species]
+		dna = new /datum/dna(null)
```
Comment (Fox-McCloud) [inline code/modules/surgery/organs/organ.dm]: this probably makes my limb grower actually work...

File: code/game/gamemodes/wizard/artefact.dm [hunk 4]
```diff
@@ -321,11 +321,11 @@ var/global/list/multiverse = list()
 	to_chat(M, "<B>You are an alternate version of [user.real_name] from another universe! Help [user.p_them()] accomplish [user.p_their()] goals at all costs.</B>")
 	M.faction = list("[user.real_name]")
 	if(duplicate_self)
-		M.set_species(user.get_species()) //duplicate the sword user's species.
+		M.set_species(user.dna.species) //duplicate the sword user's species.
```
Comment (Crazylemon64) [inline code/game/gamemodes/wizard/artefact.dm]: `species.type`

File: code/modules/mob/living/carbon/human/appearance.dm (hunk 2 above)
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/human/appearance.dm]: Is that third check sensical? Would it be better to do `!ispath(new_species, /datum/species)` here?

File: code/modules/mob/living/carbon/human/human.dm [hunk 5]
```diff
@@ -1049,11 +1038,7 @@
 	return
 
 /mob/living/carbon/human/get_species()
-
-	if(!species)
-		set_species()
-
-	return species.name
+	return dna.species.name
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/human/human.dm]: maybe I overlooked you taking care of this, but be sure to look out for places which use `get_species` as a species identifier that might be passed to something like `set_species`

File: code/modules/mob/living/carbon/human/human.dm [hunk 6]
```diff
@@ -1289,109 +1274,94 @@
...
+		skin_colour = dna.species.base_color
 	else
 		skin_colour = "#000000"
 
-	if(!(species.bodyflags & HAS_SKIN_TONE))
+	if(!(dna.species.bodyflags & HAS_SKIN_TONE))
 		s_tone = 0
 
-	species.create_organs(src)
+	dna.species.create_organs(src)
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/human/human.dm]: @Fox-McCloud The organ regeneration would be a consequence of this line here. I'm not sure how to rectify it, since it'd be difficult to tell which organs on the old species map to which organs on the new species

File: code/modules/mob/living/carbon/human/interactive/interactive.dm [hunk 7]
```diff
@@ -208,9 +208,9 @@
 			var/datum/dna/toDoppel = chosen.dna
 
 			T.real_name = toDoppel.real_name
-			T.set_species(chosen.species.name)
-			T.body_accessory = chosen.body_accessory
 			T.dna = toDoppel.Clone()
+			T.set_species(chosen.dna.species)
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/human/interactive/interactive.dm]: `species.type`

File: code/modules/vr/vr_controller.dm [hunk 8]
```diff
@@ -31,13 +31,13 @@ proc/build_virtual_avatar(mob/living/carbon/human/H, location, datum/map_templat
...
 	location = get_turf(location)
 	vr_avatar = new /mob/living/carbon/human/virtual_reality(location)
-	vr_avatar.set_species(H.species.name)
 	if(istype(H, /mob/living/carbon/human/virtual_reality))
 		var/mob/living/carbon/human/virtual_reality/V = H
 		vr_avatar.real_me = V.real_me
 	else
 		vr_avatar.real_me = H
 	vr_avatar.dna = H.dna.Clone()
+	vr_avatar.set_species(H.dna.species)
```
Comment (Crazylemon64) [inline code/modules/vr/vr_controller.dm]: `species.type`

File: code/modules/mob/living/carbon/human/human.dm (hunk 6 above)
Comment (Fox-McCloud) [reply code/modules/mob/living/carbon/human/human.dm]: I added a solution for this that returns if the species is the same as it previously was; there's an override though, or else humans, at round start, will spawn in, then immediately die and have no icons (since default DNA has a species human datum for safety).

File: code/modules/mob/living/carbon/human/human.dm (hunk 5 above)
Comment (Fox-McCloud) [reply code/modules/mob/living/carbon/human/human.dm]: I changed every single `set_species` to behave properly; this proc, AFAIK, "works", but it's not "ideal"

Ideally, we'll have a new proc that's something like `get_species_name` where a name is specifically needed, and everything else will use `is_species` (which will replaced the `if(get_species() == "Vulpkanin")` style checks.

File: code/modules/mob/living/carbon/human/human.dm (hunk 6 above)
Comment (Crazylemon64) [reply code/modules/mob/living/carbon/human/human.dm]: pfff

**Comment by Fox-McCloud** [comment]
Alright, just a few more things to add/update. Primarily abductor and claw sharpening.

After that it's testing out those things and cloning....then this can be reviewed a bit more extensively and probably tested on live.

File: code/modules/customitems/item_defines.dm [hunk 9]
```diff
@@ -45,7 +45,7 @@
 
 	var/mob/living/carbon/human/target = M
 
-	if(istype(target.species, /datum/species/machine))
+	if(istype(target.dna.species, /datum/species/machine))
```
Comment (Fox-McCloud) [inline code/modules/customitems/item_defines.dm]: Correct this.

File: code/modules/mob/living/carbon/human/emote.dm [hunk 10]
```diff
@@ -906,11 +911,18 @@
 				if("Diona")
 					emotelist += "\nDiona specific emotes :- creak(s)"
 
-			if (species.name == "Slime People")
+			if(dna.species.name == "Machine")
```
Comment (Fox-McCloud) [inline code/modules/mob/living/carbon/human/emote.dm]: use the helper

**Comment by Fox-McCloud** [comment]
And tested out cloning; it works as intended.

File: code/modules/surgery/organs/organ_external.dm [hunk 11]
```diff
@@ -105,13 +105,12 @@
...
 	..()
 	var/mob/living/carbon/human/H = holder
-	icobase = species.icobase
-	deform = species.deform
+	icobase = dna.species.icobase
+	deform = dna.species.deform
 	if(istype(H))
 		replaced(H)
 		sync_colour_to_human(H)
-	spawn(1)
```
Comment (Fox-McCloud) [inline code/modules/surgery/organs/organ_external.dm]: @KasparoVy @Crazylemon64 

Any idea why this spawn is here? I removed it, and it didn't seem to have any negative impact, other than resolving runtimes with quickly deleted organs...

File: code/modules/surgery/organs/subtypes/machine.dm [hunk 12]
```diff
@@ -110,43 +100,30 @@
...
 /obj/item/organ/internal/eyes/optical_sensor/remove(var/mob/living/user,special = 0)
 	if(!special)
 		to_chat(owner, "Error 404:Optical Sensors not found.")
 
 	. = ..()
 
 /obj/item/organ/internal/brain/mmi_holder/posibrain/New()
-	stored_mmi = new /obj/item/mmi/robotic_brain/positronic(src)
 	..()
-	spawn(1)
```
Comment (Fox-McCloud) [inline code/modules/surgery/organs/subtypes/machine.dm]: @KasparoVy @Crazylemon64 

Same here, though I doubt this even has an icon.

File: code/modules/surgery/organs/subtypes/standard.dm [hunk 13]
```diff
@@ -194,12 +194,10 @@
 			owner.unEquip(owner.r_ear)
 		if(owner.wear_mask)
 			owner.unEquip(owner.wear_mask)
-		spawn(1)
```
Comment (Fox-McCloud) [inline code/modules/surgery/organs/subtypes/standard.dm]: @KasparoVy @Crazylemon64 

Same here.

File: code/modules/mob/living/carbon/human/species/abductor.dm [hunk 14]
```diff
@@ -27,14 +27,14 @@
...
+	var/scientist = FALSE // vars to not pollute spieces list with castes
 
 /datum/species/abductor/can_understand(mob/other) //Abductors can understand everyone, but they can only speak over their mindlink to another team-member
-	return 1
+	return TRUE
 
 /datum/species/abductor/handle_post_spawn(mob/living/carbon/human/H)
 	H.gender = NEUTER
-	if(H.mind)
-		H.mind.abductor = new /datum/abductor
```
Comment (Fox-McCloud) [inline code/modules/mob/living/carbon/human/species/abductor.dm]: This means admins and now safely spawn abductors without a mind/client now and they'll function properly.

File: code/modules/mob/new_player/new_player.dm [hunk 15]
```diff
@@ -537,7 +537,7 @@
 		else if(mind.assigned_role == "Mime")
 			new_character.real_name = pick(mime_names)
 			new_character.rename_self("mime")
-		else if(new_character.species == "Diona")
+		else if(new_character.dna.species == "Diona")
```
Comment (Fox-McCloud) [inline code/modules/mob/new_player/new_player.dm]: species check

File: code/modules/surgery/organs/organ_external.dm (hunk 11 above)
Comment (KasparoVy) [reply code/modules/surgery/organs/organ_external.dm]: No ideas off the top of my head, looks to've just been waste from my point of view-- same as below

File: code/game/gamemodes/changeling/evolution_menu.dm [hunk 16]
```diff
@@ -375,7 +375,7 @@ var/list/sting_paths
 			path.on_purchase(src)
 
 	var/mob/living/carbon/C = src		//only carbons have dna now, so we have to typecaste
-	mind.changeling.absorbed_dna |= C.dna
+	mind.changeling.absorbed_dna |= C.dna.Clone()
```
[the remark this answers is not in the document]
Comment (KasparoVy) [reply code/game/gamemodes/changeling/evolution_menu.dm]: Switching to clone was the solution to a lot of old changeling related issues too. Handy

File: code/modules/awaymissions/snpc.dm [hunk 17]
```diff
@@ -4,7 +4,7 @@
 	var/squad_member = 0				// was spawned by squad
 	var/home_z
 
-/mob/living/carbon/human/interactive/away/New()
+/mob/living/carbon/human/interactive/away/Initialize(mapload)
```
Comment (Crazylemon64) [inline code/modules/awaymissions/snpc.dm]: Hmm, why this change? It's already in my maploader fix PR

File: code/modules/mob/living/carbon/human/species/plasmaman.dm [hunk 18]
```diff
@@ -6,6 +6,7 @@
 	//language = "Clatter"
 
 	species_traits = list(IS_WHITELISTED, NO_BLOOD, NOTRANSSTING)
+	skinned_type = /obj/item/stack/sheet/mineral/plasma // We're low on plasma, R&D! *eyes plasmaman co-worker intently*
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/human/species/plasmaman.dm]: pff

File: code/modules/mob/living/carbon/human/species/station.dm [hunk 19]
```diff
@@ -627,13 +604,13 @@
 #undef SLIMEPERSON_ICON_UPDATE_PERIOD
 #undef SLIMEPERSON_BLOOD_SCALING_FACTOR
 
-/mob/living/carbon/human/proc/toggle_recolor(var/silent = 0)
-	var/datum/species/slime/S = all_species[get_species()]
-	if(!istype(S))
+/mob/living/carbon/human/proc/toggle_recolor(silent = FALSE)
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/human/species/station.dm]: The `recolor_list` can probably be removed with non-global species

File: code/game/gamemodes/miniantags/abduction/abduction.dm [hunk 20]
```diff
@@ -166,11 +120,12 @@
 	H.forceMove(get_turf(L))
 	H.body_accessory = null
 	H.set_species(/datum/species/abductor)
-	H.mind.abductor.agent = 1
-	H.mind.abductor.team = team_number
+	S = H.dna.species
+	S.team = team_number
```
[the remark this answers is not in the document]
Comment (Crazylemon64) [reply code/game/gamemodes/miniantags/abduction/abduction.dm]: I *guess* it's to allow admins to spawn in mindless abductor species?? doesn't feel very comfy though

[the remark this answers is not in the document]
Comment (Fox-McCloud) [reply code/game/gamemodes/miniantags/abduction/abduction.dm]: It's not just that; it's that if we ever add in a potion to make yourself an abductor or we ever refactor genetics to transfer over the entire identity (which would include species), it seems reallyyyy messy to make an entire mind datum essentially for determining who they can talk to and if they're a scientist or not, especially when we'll generally want it to fail over player created abductors all being in the same hivemind, by default.

if it helps, think of abductors less as a deliberate antag and more as a deliberate distinct species, and this makes way more sense.

File: code/modules/awaymissions/snpc.dm (hunk 17 above)
Comment (Fox-McCloud) [reply code/modules/awaymissions/snpc.dm]: Mostly because I was getting lots of runtimes in my testing; I tried to test species wherever I could, up to, and including interactive humans--and since I had to change one of the interactive mob's anyway (to set the species to skrell, properly), it seemed a bit silly to not fix the others.

File: code/modules/mob/living/carbon/human/emote.dm [hunk 21]
```diff
@@ -29,34 +29,41 @@
 	switch(act)
 		//Cooldown-inducing emotes
 		if("ping", "pings", "buzz", "buzzes", "beep", "beeps", "yes", "no", "buzz2")
-			if(species.name == "Machine")		//Only Machines can beep, ping, and buzz, yes, no, and make a silly sad trombone noise.
+			var/found_machine_head = FALSE
+			if(ismachine(src))		//Only Machines can beep, ping, and buzz, yes, no, and make a silly sad trombone noise.
```
Comment (Crazylemon64) [inline code/modules/mob/living/carbon/human/emote.dm]: oh, you don't set `found_machine_head` here - hence IPCs not being able to use their emotes

Comment (Fox-McCloud) [reply code/modules/mob/living/carbon/human/emote.dm]: Yep. Fixed.

**Comment by Fox-McCloud** [comment]
Wooooo! I thought this day would never happen!

**Comment by KasparoVy** [comment]
Congrats, y'all did a cracking job
